### ISSUE TITLE
Domain time charged at incorrect location in MCS

### ISSUE THREAD

**Issue by xurtis** [seL4/seL4#297]
The implementation of `schedule` assumes that if the domain time would be expired that the time for the domain has already been charged before a new thread must be chosen (i.e., before the call to [`chooseNewThread`](https://github.com/seL4/seL4/blob/fd7a848b59b55a7865439028879c61dc803fc471/src/kernel/thread.c#L337-L339) in [`schedule`](https://github.com/seL4/seL4/blob/fd7a848b59b55a7865439028879c61dc803fc471/src/kernel/thread.c#L359)) otherwise a thread for the current domain will be chosen.

Under MCS, the [`commitTime`](https://github.com/seL4/seL4/blob/fd7a848b59b55a7865439028879c61dc803fc471/include/kernel/thread.h#L148-L156) function is responsible for charging the domain time and this is only called via [`switchSchedContext`](https://github.com/seL4/seL4/blob/fd7a848b59b55a7865439028879c61dc803fc471/src/kernel/thread.c#L325-L329) _after_ the next thread is chosen [`schedule`](https://github.com/seL4/seL4/blob/fd7a848b59b55a7865439028879c61dc803fc471/src/kernel/thread.c#L396).

Recommended fix is to change the check in [`chooseNewThread`](https://github.com/seL4/seL4/blob/fd7a848b59b55a7865439028879c61dc803fc471/src/kernel/thread.c#L337) to use [`isCurDomainExpired`](https://github.com/seL4/seL4/blob/fd7a848b59b55a7865439028879c61dc803fc471/include/kernel/thread.h#L118-L122) and expose that function in non-MCS configurations.

**Comment by pingerino** [comment]
How the domains and scheduling contexts interact has never clearly been defined, it specifically left it out of scope for the PhD work as I needed to finish :). It would be better to come up with and agree on a semantics first, then update & write tests for those before going forward with any changes.

**Comment by xurtis** [comment]
Yup, this is more to keep track of this particular thing not working when the two are used together (it's currently causing some issues with the proofs, though resolving these are not top priority). What may be preferable is to enable the kernel to enforce fixed-scheduled SCs in place of the domain scheduler, though for that to be useful from an infoflow perspective, they would potentially need to be gang-scheduled or be synchronisable across cores to somehow.

There were also reports of other issues with domain switching and MCS were not properly functioning in use. I'll see if I can get those added to this thread.

**Comment by pingerino** [comment]
Honestly it's probably best to make MCS + domains = invalid configuration.

**Comment by xurtis** [comment]
Yeah, I suspect the better solution is carving the domains out of the proofs (which I believe all assume the presence of domains) which would also mean dropping the infoflow proofs for MCS which I believe depend on the domain scheduler until we come up with some way of supporting them with or without it.

**Comment by lsf37** [comment]
Should look into it more to be sure, but my gut feeling is that it is not feasible to switch domains off completely for the proofs. They are mentioned explicitly on all levels of the proof, not just inflow, and it would be a major (months) operation to remove them.

It should be possible to set the number of domains to 1 explicitly and use that in the proofs. I don't think that will stop verification hitting all interaction issues, but it might reduce them.

**Comment by lsf37** [comment]
(Inflow and access control are currently not part of the proofs anyway -- what we are hitting is pure functional correctness)

**Comment by lsf37** [comment]
It looks like this is basically the same issue as in #226. Do I have that correct?

In that sense, the fix should be relatively simple -- much simpler than trying to remove the domain scheduler from the proofs. 

I do think that MCS + domain scheduler makes sense if we want MCS to be the always-on default that verification is supposed to track. When in use, that config would not get real-time guarantees from MCS any more, at least not in the traditionally analysable sense, but that is fine. You'd want those configurations for something like the CDDC, where you have fixed domains, and you whatever scheduling is available internally, not caring too much about guarantees as long as nothing starves.

That means, what we need to get right when domains are on, are not the complex real time analysis issues, but only that it makes sense for basic scheduling.

If we forbid the combination, we have no path for MCS becoming the default without giving up on infoflow, which kills a whole application area.

**Comment by lsf37** [comment]
I'm also on board with eventually replacing the domain scheduler with a concept like fixed-time SCs like @xurtis mentioned above, but we do need something for domains earlier than that.

**Comment by xurtis** [comment]
> It looks like this is basically the same issue as in #226. Do I have that correct?

I don't think so, I think they are related though. #226 deals with what is actually true at kernel exit (`ksDomainTime >= MIN_BUDGET` rather than `ksDomainTime > 0`) whereas the issue here is that the manner in which this is checked in `choosNewThread` which currently assumes that domain time is charged before that function is called when for MCS it is acutally charged after that function is called.

**Comment by tpcarpent** [comment]
> Honestly it's probably best to make MCS + domains = invalid configuration.

Ouch. I was just reading the manual (12.0.0) and in Section 6.1.1 Thread control blocks, it says, "Each thread belongs to exactly one security domain (see Section 6.3)." I was rather excited when I saw this, since my takeaway was that I could have both the MCS for multi-core and real-time flexibility, and the domains for temporal isolation to prohibit timing channels and cross-core performance attacks (which can blow real-time constraints out the window - see https://arxiv.org/abs/1903.01314 for some very bad news about modern multi-core processors).

What do you mean by "fixed-time SCs?" Would this be fixed offset and duration within a fixed rate, cyclic schedule, and synchronized as you say (or synchronizable) across cores? If so, I think something like that could be used for temporal isolation.

**Comment by xurtis** [comment]
Currently the use of more than one domain in the domain scheduler would remove any of the meaningful real-time guarantees of MCS require for real-time analysis, so you could not have both the temporal isolation (in the information-flow sense) provided by domains as well as the temporal isolation (in the real-time sense) provided by MCS at the same time.

As I currently understand, the only way a real-time system could provide the level of isolation required to prevent information flow in the scheduler is for it to enforce a fixed schedule, i.e. a task is always scheduled at the same points in the hyper-period regardless of the behavior of other threads. The requirement here is that a task can't determine if a higher-privileged task completed early by means of determining if it has executed any earlier as a result or if a higher-privileged task preempted it.

This doesn't mean that every task in the system needs to be scheduled in such a fixed manner, it just means that a higher-privileged task can't preempt a lower-privileged task on anything other than a fixed schedule and must always preempt the lower-criticality tasks for the same duration. A higher-privileged task could be preempted _by_ a lower privileged task though.

This would then require for any task that all higher-privileged tasks at an equal or higher priority are scheduled with a fixed schedule.

**Comment by xurtis** [comment]
As for multi-core processors, I'm not sure there is any way to mitigate temporal information channels without treating distinct cores as wholly distinct systems or by enforcing gang-scheduling of tasks across cores. Any optimization that utilizes slack time to allow other tasks to execute with increased budget or lower latency will also always introduce an information channel in doing so (this is even true for single-core).

**Comment by tpcarpent** [comment]
> Currently the use of more than one domain in the domain scheduler
> would remove any of the meaningful real-time guarantees of MCS require
> for real-time analysis,

If by "remove any of the meaningful real-time guarantees" you mean that the current proofs don't support it, that's fair enough. I will go back to pondering how to trick MCS into doing what I want.

From a pure scheduling perspective, I believe it is possible to hierarchically compose MCS on top of the domain scheduler and provide real-time guarantees. Davis and Burns from York did excellent hierarchical scheduling work in the 2005-2008 timeframe. For example: https://ieeexplore.ieee.org/document/1563125

> As for multi-core processors, I'm not sure there is any way to
> mitigate temporal information channels without treating distinct cores
> as wholly distinct systems 

I don't understand this statement. Modern cores aren't even remotely distinct systems. The processor vendors might want you to believe this. It is only partially true, even from a performance perspective. Certainly not true from any notion of non-interference at the micro-architectural level.

> or by enforcing gang-scheduling of tasks across cores.

I believe this is correct, given today's multi-cores. It's not how systems are currently implemented, but it will work and provide some of the desired multi-core advantages.

> Any optimization that utilizes slack time to allow other tasks to
> execute with increased budget or lower latency will also always
> introduce an information channel in doing so (this is even true for
> single-core).

I agree, and this is well understood by some. It is also acceptable - the "slack" threads wouldn't have the same security or safety criticality. There are also nuances in the term "slack time," so what you end up with depends on your definition.

> As I currently understand, the only way a real-time system could
> provide the level of isolation required to prevent information flow in
> the scheduler is for it to enforce a fixed schedule, i.e. a task is
> always scheduled at the same points in the hyper-period regardless of
> the behavior of other threads. The requirement here is that a task
> can't determine if a higher-privileged task completed early by means
> of determining if it has executed any earlier as a result or if a
> higher-privileged task preempted it.

Nicely stated.
 
> This doesn't mean that every task in the system needs to be scheduled
> in such a fixed manner, it just means that a higher-privileged task
> can't preempt a lower-privileged task on anything other than a fixed
> schedule and must always preempt the lower-criticality tasks for the
> same duration. A higher-privileged task could be preempted by a lower
> privileged task though.

Also nicely stated.
 
> This would then require for any task that all higher-privileged tasks
> at an equal or higher priority are scheduled with a fixed schedule.

"for any task with strict non-interference requirements" The nice thing is that not all tasks have this restriction, which allows us to still build high performance systems, with some acceptable overhead lost to protecting the critical sections.

**Comment by xurtis** [comment]
> If by "remove any of the meaningful real-time guarantees" you mean that the current proofs don't support it, that's fair enough. 

I mean that the set of real-time guarantees provided by the kernel in a single domain are no longer provided under multiple domains, particularly the guarantee that a thread of the highers runnable priority is currently executing as this is only the case if such a thread is in the currently selected domain. I'm not familiar with any of the scheduling algorithms that can account for a hierarchy with something like the domain scheduler as root and how they interact with the implementation of sporadic scheduler in the kernel.

**Comment by kent-mcleod** [comment]
https://github.com/seL4/seL4/pull/327 resolves this issue somewhat: When the domain time expires, a new thread from the next domain will be scheduled.  

Some further issues that are still outstanding and may need future consideration:
- Overrun isn't accounted in the same way as for SC budgets: If a thread overruns such that when it enters the kernel the domain time has been exceeded, the remaining domain time will be set to 0 and the next domain will be switched to.  The overrun amount isn't tracked anywhere so when the next iteration of the domain schedule occurs, the original domain will get a full timeslice to use again.  This means a domain that consistently overruns will have its bandwidth amplified at the expense of other domains. 
- The switch to the next domain can happen early: If the remaining time in the current domain's remaining time is less than `MIN_BUDGET` and the kernel is entered, then the current domain will be considered to have expired and the next domain will be switched to.  In this setting should the kernel be spinning for the remainder of the domain's timeslice?
- The domain schedule can drift due to the lack of synchronization between domain activation time and the clock time.  As there will often be a small error between the current time and the proper end time of the current domain, the next domain is going to be started slightly early or late relative to the previous domains start time.  This difference can then accumulate over multiple periods.
